### PR TITLE
Fix php::fpm eternal reload for mysqli a simplexml extension

### LINKED ISSUE #309: Ubuntu 16 mysql extension so name with packages from ondrej PPA
Using Ubuntu 16.04 LTS with the PHP builds from https://launchpad.net/~ondrej/+archive/ubuntu/php, the mysql extension installs the so files mysqli, mysqlnd and pdo_mysql.

The Puppet module automatically creates an extension ini file named after the extension itself. However mysql.so does not exist. Whilst an option is available to change the so name, not one but three so files are installed.

## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 4.9.0
- Ruby: ruby 2.1.9p490 (2016-03-30 revision 54437) [x86_64-linux]
- Distribution: Ubuntu 16.04 LTS
- Module version: 4.0.0

## How to reproduce (e.g Puppet code you use)
Add the extension for mysql. I use hiera to achieve this:

php::extensions:
  mysql: {}

## What are you seeing
The puppet module initialises a php extension ini file for the mysql extension, however the extension contains 3 so files for mysqli, mysqlnd and pdo_mysql. Whilst the module has support to override the so name, it can't override all 3.

## What behaviour did you expect instead
The ability to disable generating an ini file in the puppet extension manifest. Optionally allowing custom ini files to be dropped in to place using the php::config manifest via the main init.pp script and via hiera.

### LINKED ISSUE #497: Extension mysqli causes php7-fpm to reload
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.5.8
- Ruby:
- Distribution: Ubuntu 16.04
- Module version: 6.0.2

## How to reproduce (e.g Puppet code you use)
class profiles::php7 {
# include ::apt because of a bug in php-module.
  include ::apt

# We need  at least php7.0 php7.0-fpm php7.0-gd php7.0-xml php7.0-mbstring libgd-dev composer.

  class { '::php::globals':
    php_version => '7.0',
  }->
  class { '::php':
    ensure       => 'present',
    manage_repos => false,
    fpm          => true,
    dev          => true,
    composer     => true,
    pear         => true,
    phpunit      => false,
    fpm_user  => 'www-data',
    fpm_group => 'www-data',
    #require => [User['www-data']],
    extensions => {
      mbstring => { },
      curl => { },
      gd => { },
      xmlrpc => { },
      bcmath => { },
      zip => { },
      mysqli => { },
      # redis => { },
      mcrypt => { },
      intl => { },
      soap => { },
      sqlite3 => { },
    },
    #  settings   => {
    #    'PHP/max_execution_time'  => '30',
    #  },
  }
  php::fpm::pool { 'common':
    # http://php.puppet.mayflower.de/php/fpm/pool.html
    ensure => 'present',
    listen => '/var/run/php-fpm-common.sock',
    listen_owner => 'www-data',
    listen_group => 'www-data',
  }
}

## What are you seeing
php7-fpm is triggered to restart every time puppet agent is run.
The restart is triggered by installation of extension mysqli for every puppet agent run.
No other extension i am using are causing the same issue.
## What behaviour did you expect instead
Installation of modul once.
## Output log
Notice: /Stage[main]/Php/Php::Extension[mysqli]/Php::Extension::Install[mysqli]/Package[php7.0-mysqli]/ensure: created
Info: Php::Extension::Install[mysqli]: Scheduling refresh of Php::Extension::Config[mysqli]
Info: Php::Extension::Config[mysqli]: Scheduling refresh of Php::Config[mysqli]
Info: Php::Extension::Config[mysqli]: Scheduling refresh of Exec[/usr/sbin/phpenmod -v 7.0 -s ALL mysqli]
Info: Php::Config[mysqli]: Scheduling refresh of Php::Config::Setting[/etc/php/7.0/mods-available/mysqli.ini: extension]
Info: Php::Config::Setting[/etc/php/7.0/mods-available/mysqli.ini: extension]: Scheduling refresh of Ini_setting[/etc/php/7.0/mods-available/mysqli.ini: extension]
Notice: /Stage[main]/Php/Php::Extension[mysqli]/Php::Extension::Config[mysqli]/Php::Config[mysqli]/Php::Config::Setting[/etc/php/7.0/mods-available/mysqli.ini: extension]/Ini_setting[/etc/php/7.0/mods-available/mysqli.ini: extension]: Triggered 'refresh' from 1 event
Notice: /Stage[main]/Php/Php::Extension[mysqli]/Php::Extension::Config[mysqli]/Exec[/usr/sbin/phpenmod -v 7.0 -s ALL mysqli]: Triggered 'refresh' from 1 event
Info: Php::Extension[mysqli]: Scheduling refresh of Service[php7.0-fpm]
Notice: /Stage[main]/Php::Fpm::Service/Service[php7.0-fpm]: Triggered 'refresh' from 1 event
Notice: Applied catalog in 3.47 seconds

### PR DESCRIPTION
**Pull request by miranovy**
#### Pull Request (PR) description
Fix eternal reload for mysqli and simplexml extension. 

Package name for mysqli extension is php7.2-mysql. Package name for simlexml extension is php7.2-xml. I can use package_prefix variable, than i add package_name variable with higher priority.

Example of configuration:
```puppet
class { '::php':
        ensure       => latest,
        manage_repos => false,
        pear         => true,
        composer     => false,
        fpm          => false,
        dev          => false,
        extensions   => {
            bcmath    => {},
            curl      => {},
            gd        => {},
            mbstring  => {},
            simplexml => {
                package_name => 'php7.2-xml',
            },
            mysqli    => {
                package_name => 'php7.2-mysql',
            },
            sqlite3   => {},
        },
    }
```

#### This Pull Request (PR) fixes the following issues
Fixes #309 
Fixes #497

### CODE REVIEW

File: manifests/extension.pp [hunk 1]
```diff
@@ -74,6 +78,7 @@
   Optional[String] $ini_prefix                      = undef,
   Optional[String] $php_api_version                 = undef,
   String           $package_prefix                  = $php::package_prefix,
+  Optional[String] $package_name                    = undef,
```
Comment (bastelfreak) [inline manifests/extension.pp]: please enforce the minimal string length:
```suggestion
  Optional[String[1]] $package_name                 = undef,
```

File: manifests/extension/install.pp [hunk 2]
```diff
@@ -38,6 +42,7 @@
   Optional[Php::Provider] $provider                 = undef,
   Optional[String] $source                          = undef,
   String           $package_prefix                  = $php::package_prefix,
+  Optional[String] $package_name                    = undef,
```
Comment (bastelfreak) [inline manifests/extension/install.pp]:
```suggestion
  Optional[String[1]] $package_name                 = undef,
```

**Comment by bastelfreak** [comment]
Hi @miranovy, thank for the bugfix. Can you add an acceptance test for this? We have some examples in https://github.com/voxpupuli/puppet-php/tree/master/spec/acceptance. Let us know if you need some help! We're available in our IRC channel #voxpupuli on freenode and at slack https://slack.puppet.com

File: manifests/extension.pp (hunk 1 above)
Comment (miranovy) [reply manifests/extension.pp]: Added

**Comment by miranovy** [comment]
Hi @bastelfreak, you mean something like that in `phpspec.rb`:
```rb
...

 case default[:platform]
    when %r{ubuntu-18.04}
      packagename = 'php7.2-fpm'
      simplexmlpackagename = 'php7.2-xml'
    when %r{ubuntu-16.04}
      packagename = 'php7.0-fpm'
      simplexmlpackagename = 'php7.0-xml'
    when %r{ubuntu-14.04}
      packagename = 'php5-fpm'
      simplexmlpackagename = 'php5-xml'
    when %r{el}
      packagename = 'php-fpm'
    when %r{debian-8}
      packagename = 'php5-fpm'
    when %r{debian-9}
      packagename = 'php7.0-fpm'
    end
    describe package(packagename) do
      it { is_expected.to be_installed }
    end

    describe service(packagename) do
      it { is_expected.to be_running }
      it { is_expected.to be_enabled }
    end
  end
  context 'default parameters with extensions' do
    case default[:platform]
    when %r{ubuntu-18.04}, %r{ubuntu-16.04}, %r{ubuntu-14.04}
      it 'works with defaults' do
        pp = <<-EOS
        class{'php':
          extensions => {
            'mysql'    => {},
            'gd'       => {},
            'net-url'  => {
              package_prefix => 'php-',
              settings       => {
                extension => undef
              },
            'simplexml'  => {
              package_name => simplexmlpackagename
            }
          }
        }
        EOS
        # Run it twice and test for idempotency
        apply_manifest(pp, catch_failures: true)
        apply_manifest(pp, catch_changes: true)
      end
    else
      it 'works with defaults' do
        pp = <<-EOS
        class{'php':
          extensions => {
            'mysql'    => {},
            'gd'       => {}
          }
        }
        EOS
        # Run it twice and test for idempotency
        apply_manifest(pp, catch_failures: true)
        apply_manifest(pp, catch_changes: true)
      end
    end
...
```

**Comment by bastelfreak** [comment]
Hi,
yes that looks good.

**Comment by bastelfreak** [comment]
@miranovy can you take a look at the failed travis jobs ?

File: spec/acceptance/php_spec.rb [hunk 3]
```diff
@@ -45,7 +53,10 @@
               package_prefix => 'php-',
               settings       => {
                 extension => undef
-              }
+              },
+            }
+            'simplexml'  => {
+              package_name => simplexmlpackagename,
```
Comment (bastelfreak) [inline spec/acceptance/php_spec.rb]: the value of a hash key needs to be quoted. In this case:
```suggestion
              package_name => 'simplexmlpackagename',
```

But I don't think that `simplexmlpackagename` is a valid name for any package on any distribution?

Comment (miranovy) [reply spec/acceptance/php_spec.rb]: Hi,
  package name for simplexml depends on used OS (php7.2-xml for ubuntu 1804, php7.0-xml for ubuntu 1604 and php-xml for ubutu 1404).

I am trying set the correct name by switch (see https://github.com/voxpupuli/puppet-php/pull/503/files#diff-2e2ff95fcba379972e19088c70dd68a5R39 ), but it not work. Can you help me with this, please?

Comment (bastelfreak) [reply spec/acceptance/php_spec.rb]: ahh my bad. This should work:
```suggestion
              package_name => '#{simplexmlpackagename}',
```

File: spec/acceptance/php_spec.rb [hunk 4]
```diff
@@ -45,7 +53,10 @@
               package_prefix => 'php-',
               settings       => {
                 extension => undef
-              }
+              },
+            }
```
Comment (bastelfreak) [inline spec/acceptance/php_spec.rb]: There is a comma missing:
```suggestion
            },
```

**Comment by bastelfreak** [comment]
Can you take a look at the failing travis jobs?

**Comment by miranovy** [comment]
> Can you take a look at the failing travis jobs?

There was duplicity with default xml package for ubuntu (file manifests/dev.pp:47).

**Comment by bastelfreak** [comment]
Thanks!

**Comment by miranovy** [comment]
> Thanks!

You're welcome. Thank you for merge.

**Pull request merged**